src/Object/ObjectFactory.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+/*
+ * This file is part of the zenstruck/foundry package.
+ *
+ * (c) Kevin Bond <kevinbond@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Zenstruck\Foundry\Object;
+
+use Zenstruck\Foundry\BaseFactory;
+use Zenstruck\Foundry\FactoryCollection;
+use Zenstruck\Foundry\Instantiator;
+use Zenstruck\Foundry\LazyValue;
+
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ *
+ * @template T of object
+ * @extends BaseFactory<T>
+ *
+ * @phpstan-type CallableInstantiator Instantiator<T>|\Closure(Parameters,class-string<T>):T
+ */
+abstract class ObjectFactory extends BaseFactory
+{
+    /** @var CallableInstantiator */
+    private \Closure|Instantiator $instantiator;
+
+    /** @var array<callable(Parameters,class-string<T>):Parameters> */
+    private array $beforeInstantiate = [];
+
+    /** @var array<callable(T,Parameters):void> */
+    private array $afterInstantiate = [];
+
+    /**
+     * @return class-string<T>
+     */
+    abstract protected static function getClass(): string;
```

Comment:
For BC sake, I kept the same name, and did not changed for `class()`, even if I do prefer this name